### ISSUE TITLE
Picasso 2.5 can't add auth header

### ISSUE THREAD

**Issue by favn1585** [square/picasso#900]
At Picasso 2.4 we were added auth token using

``` java
        OkHttpDownloader downloader = new OkHttpDownloader(context) {
            @Override
            protected HttpURLConnection openConnection(Uri uri) throws IOException {
                HttpURLConnection connection = super.openConnection(uri);
                connection.setRequestProperty(RestClient.AUTHORIZATION_KEY, RestClient.getValidAuth());
                return connection;
            }
        };
```

At 2.5 method

```java
super.openConnection(uri);
```
not available. 
Also if we use UrlConnectionDownloader images are not cached to disc.

**Comment by JakeWharton** [comment]
If you are using OkHttp with Picasso, the right way to add headers is with an [interceptor](https://github.com/square/okhttp/wiki/Interceptors). This will allow modifying every request that is made by Picasso through OkHttp without having to mess around with subclassing Picasso classes.

If you re-use your `OkHttpClient` instance and only want the headers added for images, call `.clone()` on the client before adding the interceptor and handing it to Picasso. This will share all of the underlying bits like caching and connection pooling with the original instance but only add the headers for Picasso requests.
```java
OkHttpClient picassoClient = client.clone();
picassoClient.interceptors().add(new MyInterceptor());
// ...
new Picasso.Builder(context).downloader(new OkHttpDownloader(picassoClient)).build();
```

**Comment by bryant1410** [comment]
But this way responses won't be cached, will they? See  https://github.com/square/picasso/blob/eb2e9730fb7dbe25b4ab9d4ba5d2050c70c27024/picasso/src/main/java/com/squareup/picasso/OkHttpDownloader.java#L84

**Comment by lewisevans** [comment]
@bryant1410 Wouldn't this work? 
``` java
File httpCacheDir = new File(ApplicationUtils.getApplicationContext().getExternalCacheDir(), "whatever");
long httpCacheSize = 10 * 1024 * 1024; // 10 MiB
try {
    Cache cache = new Cache(httpCacheDir, httpCacheSize);
    client.setCache(cache);
} catch (IOException e) {
    e.printStackTrace();
}
```

**Comment by bryant1410** [comment]
Yep you are right :smile:

**Comment by alioth865** [comment]
Hi, i've been trying with this
```Java
public class Myinterceptor implements Interceptor {
    private String basicAuth;

    public MyInterceptor(String usuario, String contraseña) {
        String userpass= usuario + ":" + contraseña;
        basicAuth = "Basic " + new String(Base64.encodeBase64(userpass.getBytes()));
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request newRequest = chain.request().newBuilder()
                .addHeader("Authorization", basicAuth)
                .build();
        //DUDA NUNCA PASA POR AQUI;
        Log.i("Peticion", newRequest.toString());
        Log.i("Cabecera", newRequest.headers().toString());
        return chain.proceed(newRequest);
    }
}
```

but still doesn't show me the image, the way i use is:

```Java
        String urlImagenDefecto="172.19.216.87/prestashop/api/images/products/1/1";
        String key="88IWBKN2CG5EENND8MTVA4EKD57RN8X6";
        OkHttpClient picassoClient = new OkHttpClient();
        picassoClient.interceptors().add(new MyInterceptor(key,""));
        Picasso picasso=new Picasso.Builder(actividad).downloader(new OkHttpDownloader(picassoClient)).build();
        picasso.load(urlImagenDefecto).into(holder.imagen);
```

**Comment by dvdrlee** [comment]
An addendum to @JakeWharton's comment above: you'll want to make sure you are on v2.4.0 of OkHttp.  I was on 2.3.0 and it turns out the interceptors were shared between clones, this meant that the header was being added to all my requests. Looks like it was fixed in the 2.4.  Details in the changelog here https://github.com/square/okhttp/blob/master/CHANGELOG.md

**Comment by max-pfeiffer** [comment]
Another addendum to @JakeWharton's comment above: when using OkHttp3 you want to use Jake's [OkHttp3Downloader](https://github.com/JakeWharton/picasso2-okhttp3-downloader).
